modules/gdpr_tasks/src/Anonymizer.php

```diff
@@ -124,73 +97,24 @@ public function run(TaskInterface $task) {
     // Make sure we load a fresh copy of the entity (bypassing the cache)
     // so we don't end up affecting any other references to the entity.
     $user = $task->getOwner();
-
     $errors = [];
-    $entities = [];
-    $successes = [];
-    $failures = [];
-    $log = [];
 
     if (!$this->checkExportDirectoryExists()) {
       $errors[] = 'An export directory has not been set. Please set this under Configuration -> GDPR -> Right to be Forgotten';
+      return $errors;
     }
 
-    $this->collector->getValueEntities($entities, 'user', $user);
-
-    foreach ($entities as $bundles) {
-      foreach ($bundles as $bundle_entity) {
-        // Re-load a fresh copy of the bundle entity from storage so we don't
-        // end up modifying any other references to the entity in memory.
-        $bundle_entity = $this->entityTypeManager->getStorage($bundle_entity->getEntityTypeId())
-          ->loadUnchanged($bundle_entity->id());
-
-        $entity_success = TRUE;
-
-        foreach ($this->getFieldsToProcess($bundle_entity) as $field_info) {
-          /** @var \Drupal\Core\Field\FieldItemListInterface $field */
-          $field = $field_info['field'];
-          $mode = $field_info['mode'];
-
-          $success = TRUE;
-          $msg = NULL;
-          $anonymizer = '';
+    // Traverser does the actual anonymizing.
+    $result = $this->traversal->traverse($user);
 
-          if ($mode == 'anonymize') {
-            list($success, $msg, $anonymizer) = $this->anonymize($field, $bundle_entity);
-          }
-          elseif ($mode == 'remove') {
-            list($success, $msg) = $this->remove($field);
-          }
-
-          if ($success === TRUE) {
-            $log[] = [
-              'entity_id' => $bundle_entity->id(),
-              'entity_type' => $bundle_entity->getEntityTypeId() . '.' . $bundle_entity->bundle(),
-              'field_name' => $field->getName(),
-              'action' => $mode,
-              'anonymizer' => $anonymizer,
-            ];
-          }
-          else {
-            // Could not anonymize/remove field. Record to errors list.
-            // Prevent entity from being saved.
-            $entity_success = FALSE;
-            $errors[] = $msg;
-          }
-        }
-
-        if ($entity_success) {
-          $successes[] = $bundle_entity;
-        }
-        else {
-          $failures[] = $bundle_entity;
-        }
-      }
-    }
+    $log = $result['log'];
+    $errors = $result['errors'];
+    $successes = $result['successes'];
```

Comment:
We probably need to also add in a `$result['deletions']` so they can happen inside the transaction.

Reply:
Done